cmd/run/cmd.go

```diff
@@ -0,0 +1,332 @@
+			Options:     []string{"ro"},
+			Type:        define.TypeBind,
+		}, {
+			Source:      homeDir + "/.config/gcloud/credentials.db",
+			Destination: "/root/.config/gcloud/credentials_readonly.db",
+			Options:     []string{"ro"},
+			Type:        define.TypeBind,
+		}, {
+			Source:      homeDir + "/.config/gcloud/access_tokens.db",
+			Destination: "/root/.config/gcloud/access_tokens_readonly.db",
+			Options:     []string{"ro"},
+			Type:        define.TypeBind,
+		},
+	}
+}
+
+func awsCredentialsMounts() []specs.Mount {
+	homeDir, _ := os.UserHomeDir()
+	return []specs.Mount{
+		{
+			Source:      homeDir + "/.aws/credentials",
+			Destination: "/root/.aws/credentials",
+			Options:     []string{"ro"},
+			Type:        define.TypeBind,
+		},
+		{
+			Source:      homeDir + "/.aws/config",
+			Destination: "/root/.aws/config",
+			Options:     []string{"ro"},
+			Type:        define.TypeBind,
+		},
+	}
+}
+
+func macAgentLocation() (string, error) {
+	dirs, err := os.ReadDir("/private/tmp")
+	if err != nil {
+		return "", err
+	}
+
+	if len(dirs) < 1 {
+		return "", errors.New("no dirs found at /private/tmp")
+	}
+
+	for _, dir := range dirs {
+		if dirName := dir.Name(); strings.Contains(dirName, "com.apple.launchd") {
+			return dirName, nil
+		}
+	}
+
+	return "", errors.New("no dir found at /private/tmp containing com.apple.launchd")
+}
+
+func copyPortmap(conn context.Context, containerId string) {
+	tmpdir, err := os.MkdirTemp("", "occ_portmaps")
+	if err != nil {
+		log.Fatal("Failed to create a tempdir for portmap")
+	}
+	defer os.RemoveAll(tmpdir)
+
+	data, err := containers.Inspect(conn, containerId, nil)
+	if err != nil {
+		log.Fatal("Failed to inspect container")
+	}
+
+	var hostPorts []string
+	for _, host := range data.NetworkSettings.Ports["9999/tcp"] {
+		hostPorts = append(hostPorts, host.HostPort)
+	}
+
+	portmapFile, err := os.Create(tmpdir + "/portmap")
+	if err != nil {
+		log.Fatal("Failed to create portmap file")
+	}
+
+	for _, port := range hostPorts {
+		_, err := fmt.Fprintln(portmapFile, port)
+		if err != nil {
+			log.Fatal("Failed to write host port to portmap file")
+		}
+	}
+
+	reader, writer := io.Pipe()
+	hostCopy := func() error {
+		defer writer.Close()
+		getOptions := buildahCopiah.GetOptions{
+			KeepDirectoryNames: true,
+		}
+		if err := buildahCopiah.Get("/", "", getOptions, []string{portmapFile.Name()}, writer); err != nil {
+			return fmt.Errorf("error copying portmap file from host")
+		}
+		return nil
+	}
+
+	containerCopy := func() error {
+		defer reader.Close()
+		copyFunc, err := containers.CopyFromArchive(conn, containerId, "/tmp", reader)
+		if err != nil {
+			return err
+		}
+		if err := copyFunc(); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	if err := doCopy(hostCopy, containerCopy); err != nil {
+		log.Fatal("Error copying portmap file to host.")
+	}
+}
+
+func doCopy(hostCopyFunc func() error, containerCopyFunc func() error) error {
```

Comment:
Can we add a comment here as to why we're doing this the way it's being done?  At first glance my brain is going "oh, weird, why are we using channels."  While I know there's a very good reason here, I fear this function becoming a black box that nobody will dare touch in time if we don't add comments here.